src/ln/channelmanager.rs

```diff
@@ -667,6 +675,172 @@ impl ChannelManager {
+					short_channel_id: 0,
+					prev_short_channel_id: 0,
+					amt_to_forward: next_hop_data.data.amt_to_forward,
+					outgoing_cltv_value: next_hop_data.data.outgoing_cltv_value,
+				})
+			} else {
+				let mut new_packet_data = [0; 20*65];
+				chacha.process(&msg.onion_routing_packet.hop_data[65..], &mut new_packet_data[0..19*65]);
+				chacha.process(&ChannelManager::ZERO[0..65], &mut new_packet_data[19*65..]);
+
+				let mut new_pubkey = msg.onion_routing_packet.public_key.clone();
+
+				let blinding_factor = {
+					let mut sha = Sha256::new();
+					sha.input(&new_pubkey.serialize()[..]);
+					sha.input(&shared_secret[..]);
+					let mut res = [0u8; 32];
+					sha.result(&mut res);
+					match SecretKey::from_slice(&self.secp_ctx, &res) {
+						Err(_) => {
+							// Return temporary node failure as its technically our issue, not the
+							// channel's issue.
+							return_err!("Blinding factor is an invalid private key", 0x2000 | 2, &[0;0]);
+						},
+						Ok(key) => key
+					}
+				};
+
+				match new_pubkey.mul_assign(&self.secp_ctx, &blinding_factor) {
+					Err(_) => {
+						// Return temporary node failure as its technically our issue, not the
+						// channel's issue.
+						return_err!("New blinding factor is an invalid private key", 0x2000 | 2, &[0;0]);
+					},
+					Ok(_) => {}
+				};
+
+				let outgoing_packet = msgs::OnionPacket {
+					version: 0,
+					public_key: new_pubkey,
+					hop_data: new_packet_data,
+					hmac: next_hop_data.hmac.clone(),
+				};
+
+				//TODO: Check amt_to_forward and outgoing_cltv_value are within acceptable ranges!
+
+				PendingHTLCStatus::Forward(PendingForwardHTLCInfo {
+					onion_packet: Some(outgoing_packet),
+					payment_hash: msg.payment_hash.clone(),
+					short_channel_id: next_hop_data.data.short_channel_id,
+					prev_short_channel_id: 0,
+					amt_to_forward: next_hop_data.data.amt_to_forward,
+					outgoing_cltv_value: next_hop_data.data.outgoing_cltv_value,
+				})
+			};
+
+		channel_state = Some(self.channel_state.lock().unwrap());
+		if let &PendingHTLCStatus::Forward(PendingForwardHTLCInfo { ref onion_packet, ref short_channel_id, .. }) = &pending_forward_info {
+			if onion_packet.is_some() { // If short_channel_id is 0 here, we'll reject them in the body here
+				let id_option = channel_state.as_ref().unwrap().short_to_id.get(&short_channel_id).cloned();
+				let forwarding_id = match id_option {
+					None => {
+						return_err!("Don't have available channel for forwarding as requested.", 0x4000 | 10, &[0;0]);
+					},
+					Some(id) => id.clone(),
+				};
+				if let Some(chan_update) = {
+					let chan = channel_state.as_mut().unwrap().by_id.get_mut(&forwarding_id).unwrap();
+					if !chan.is_live() {
+						Some(self.get_channel_update(chan).unwrap())
+					} else { None }
+				} {
+					return_err!("Forwarding channel is not in a ready state.", 0x1000 | 7, &chan_update.encode_with_len()[..]);
+				}
+			}
+		} else { unreachable!(); }
```

Comment:
Seems useless else branch for me, I build without it?

Reply:
Fixed.